python/lsst/pipe/base/pipelineIR.py

```diff
@@ -874,17 +946,12 @@ def from_uri(cls, uri: Union[str, ButlerURI]) -> PipelineIR:
             The loaded pipeline
         """
         loaded_uri = ButlerURI(uri)
-        # With ButlerURI we have the choice of always using a local file or
-        # reading in the bytes directly. Reading in bytes can be more
-        # efficient for reasonably-sized files when the resource is remote.
-        # For now use the local file variant. For a local file as_local() does
-        # nothing.
-        with loaded_uri.as_local() as local:
+        with loaded_uri.open("r") as buffer:
             # explicitly read here, there was some issue with yaml trying
             # to read the ButlerURI itself (I think because it only
             # pretends to be conformant to the io api)
-            loaded_yaml = yaml.load(local.read(), Loader=PipelineYamlLoader)
-            return cls(loaded_yaml)
+            loaded_yaml = yaml.load(buffer, Loader=PipelineYamlLoader)
+            return cls(loaded_yaml, directory=loaded_uri.parent())
```

Comment:
Since we need this behavior for file-like URIs (including e.g. S3, where we simulate directories with path manipulations), it sounds like we just need to trap for the case where this won't work, and raise a friendlier exception.  Are you saying that we can do that by seeing if `loaded_uri.parent() == loaded_uri`?  If not, I'm not sure how best to check, and I'm hoping you (or @timj) might.

Reply:
I'm trying to reconstruct my thinking on this from before Christmas (the exact fall over mechanism I noticed) but it made be wonder another thing. If we are saying a finalized pipeline may be a directory, then this method just taking a URI would fall over if someone ever tried to pass that URI to this method.